### PR TITLE
Add support for empty ports

### PR DESCRIPTION
**Pull request by ItsKev**
**What type of PR is this?**
> Uncomment only one ` /kind <>` line, hit enter to put that in a new line, and remove leading whitespace from that line:
>
> /kind breaking
> /kind bug
> /kind cleanup
> /kind documentation
> 
/kind feature
> /kind hotfix

**What this PR does / Why we need it**:

This PR adds support for empty ports.

**Which issue(s) this PR fixes**:

Closes https://github.com/googleforgames/agones/issues/749

**Special notes for your reviewer**:

This is my first time working with GO. If I can improve anything feel free to point me in the right direction.
Somehow the install.yaml shows a lot of changes. Looking through them in GoLand however, does not show me any changes at all on certain lines.

### CODE REVIEW

**Comment by markmandel** [comment]
Looks like Helm is being grumpy with the change to the CRD :grimacing: it's been doing that as of late with the new schema changes, I can handhold that one through e2e testing.

At first glance, this looks good (I'll do a deeper review at some point) - the only thing I would advocate for - somewhere in https://github.com/googleforgames/agones/blob/main/test/e2e/gameserver_test.go write an e2e test that confirms you can create a GameServer with no ports, and nothing falls over.

Thanks for doing this work!

**Comment by markmandel** [comment]
I'm getting this weird error I get sometimes with Github

image

So I'll wait for the e2e test to come in and walk it through the e2e tests then.

If you run into issues / would like some help, please feel free to drop into #development on Slack, and ask questions as well!

**Comment by ItsKev** [comment]
Gonna write the e2e test tomorrow 👍

**Comment by ItsKev** [comment]
I've added an e2e test but was not able to test it locally against a Google Cloud cluster. I get the same error while trying to use `make install` that happens in the tests here.

**Comment by markmandel** [comment]
This looks good, looks like you are hitting https://github.com/helm/helm/issues/5806 which won't get fixed until K8s 1.20 :confused: We're hitting this in a couple of other places.

I'll hand hold this through e2e test shortly.

**Comment by markmandel** [comment]
Hey - can you give me permission to update your branch please? I can't update it to master.

image

**Comment by ItsKev** [comment]
The allow edits by maintainers is not possible for organization repositories.

@markmandel I've invited you to write to the repository. If this is not how you handle things I could fork the project with my private account and create a PR again.

**Comment by markmandel** [comment]
> The allow edits by maintainers is not possible for organization repositories.

AAAAAAAAAAAAAAAAAAAAAAAAAAAH! That's why I can't update SOME PR's!!! This has been driving me up the wall for eons.

Don't give me write permissions on your repo -- that's seems icky.

Ping me on Slack when you want to update this PR to master, and I can handhold it through e2e tests :+1:

**Comment by markmandel** [comment]
Actually, if you are happy, I'll take the write permissions on the repo -- let me know whichever way works best for you.

**Comment by markmandel** [comment]
Can you run `make gen-install` again please? Your new changes combined with the recent updates to the CRDs because of #2007  aren't in that file. Thanks!

**Comment by markmandel** [comment]
Just a heads up - RC is next Tuesday, but let's see if we can squeeze this in.

File: test/e2e/framework/framework.go
Comment (markmandel) [inline test/e2e/framework/framework.go:217]: I'm trying to decide if `CreateGameServerAndWaitUntilReadyWithOptionalPortCheck` is necessary.

We already cover this by checking on ports in [this test](https://github.com/googleforgames/agones/blob/170384f993e4919d31950e5f995924de22fec8ef/test/e2e/gameserver_test.go#L54) for example (I'm sure there are others).

@roberthbailey WDYT? 

I get the impetus to check all other tests, but I'm not sure it's required.

File: test/e2e/gameserver_test.go
Comment (markmandel) [inline test/e2e/gameserver_test.go:781]: Nit:

This `if err != nil` block can be changed to:
`require.NoError(err)`  (I know we have the `assert.FailNow()` pattern in a bunch of places, but that was before we had access to `require` :smile: )

File: test/e2e/gameserver_test.go
Comment (markmandel) [inline test/e2e/gameserver_test.go:785]: Just a thought. This might be better, just because nil vs an empty array are both valid return results. Not a strong opinion.

https://pkg.go.dev/github.com/stretchr/testify/assert#Empty

```suggestion
	assert.Empty(t, readyGs.Spec.Ports) 
```

**Review by markmandel — COMMENTED** [review]
Yay we passed!

Just had a couple of minor things that I think could be cleaned up. Would also like a second opinion on `CreateGameServerAndWaitUntilReadyWithOptionalPortCheck` :thinking:

File: test/e2e/gameserver_test.go
Comment (ItsKev) [reply test/e2e/gameserver_test.go]: Done

File: test/e2e/gameserver_test.go
Comment (ItsKev) [reply test/e2e/gameserver_test.go]: Yes, you are right. assert.Empty itself does a null check.

Done

File: test/e2e/framework/framework.go
Comment (roberthbailey) [reply test/e2e/framework/framework.go:217]: Looking at your new test code, I'm guessing that you added this because without passing in false, line 227 (in the old file) was returning an error, because there was no port. 

But the test should be able to inspect the passed in game server and tell if a port should be created (by seeing if spec.Ports is nil). And then you can modify the check for ports to read something like "if the spec has a port and none was allocated, return an error" instead of what we have currently ("assume that there will always be a port so check to see if one was allocated").

Comment (markmandel) [reply test/e2e/framework/framework.go:217]: https://github.com/googleforgames/agones/blob/main/test/e2e/framework/framework.go#L227 - just for reference :smile: 

That's a really good point @roberthbailey , that could become:

```go
	if len(readyGs.Status.Ports) == len(gsCopy.Spec.Ports) ​{
		return nil, fmt.Errorf("Ready GameServer instance has no port: %v", readyGs.Status)
	}
```

Don't think we need nil checks either, since `len(nil)` is `0`

Which is far more elegant :+1:

Comment (roberthbailey) [reply test/e2e/framework/framework.go:217]: I think you mean != 

Also, the error message needs to be updated to print the expected number of ports and the observed number of ports:


```go
	if len(readyGs.Status.Ports) != len(gsCopy.Spec.Ports) ​{
		return nil, fmt.Errorf("Ready GameServer instance has %d port(s), want %v", readyGs.Status, gsCopy.Spec.Ports)
	}
```

Comment (markmandel) [reply test/e2e/framework/framework.go:217]: Ah yes! I did, excellent catch. I like it!

Comment (ItsKev) [reply test/e2e/framework/framework.go]: Thanks for the input. I've updated the PR accordingly. 👍

**Comment by ItsKev** [comment]
The PR got reviewed and is ready to be merged if my changes are accepted.

/assign @cyriltovena

**Review by markmandel — APPROVED** [review]
I like it! Let's get this merged!

**Pull request merged**
